src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandResultBuilder.java

```diff
@@ -0,0 +1,142 @@
+  private final boolean useErrorObjectV2;
+
+  // Created in the Ctor
+  private final Function<APIException, CommandResult.Error> apiExceptionToError;
+
+  public CommandResultBuilder(
+      ResponseType responseType, boolean useErrorObjectV2, boolean debugMode) {
+    this.responseType = responseType;
+    this.useErrorObjectV2 = useErrorObjectV2;
+    this.debugMode = debugMode;
+
+    this.apiExceptionToError = new APIExceptionCommandErrorBuilder(debugMode, useErrorObjectV2);
+  }
+
+  public CommandResultBuilder addStatus(CommandStatus status, Object value) {
+    cmdStatus.put(status, value);
+    return this;
+  }
+
+  public CommandResultBuilder addThrowable(Throwable thorwable) {
+    var error = throwableToCommandError(thorwable);
+    return addCommandError(error);
+  }
+
+  public CommandResultBuilder addCommandError(CommandResult.Error error) {
+    cmdErrors.add(error);
+    return this;
+  }
+
+  public CommandResultBuilder addDocument(JsonNode document) {
+    documents.add(document);
+    return this;
+  }
+
+  public CommandResultBuilder addDocuments(List<JsonNode> documents) {
+    this.documents.addAll(documents);
+    return this;
+  }
+
+  public CommandResultBuilder addWarning(String warning) {
+    warnings.add(warning);
+    return this;
+  }
+
+  public CommandResultBuilder nextPageState(String nextPageState) {
+    this.nextPageState = nextPageState;
+    return this;
+  }
+
+  /**
+   * Gets the appropriately formatted error given {@link #useErrorObjectV2} and {@link #debugMode}.
+   */
+  public CommandResult.Error throwableToCommandError(Throwable throwable) {
+
+    if (throwable instanceof APIException apiException) {
+      // new v2 error object, with family etc.
+      // the builder will handle the debug mode and extended errors settings to return a V1 or V2
+      // error
+      return apiExceptionToError.apply(apiException);
+    }
+
+    // the mapper handles error object v2 part
+    return ThrowableToErrorMapper.getMapperWithMessageFunction()
+        .apply(throwable, throwable.getMessage());
+  }
+
+  public CommandResult build() {
+
+    switch (responseType) {
+      case SINGLE_DOCUMENT:
+        if (documents.size() > 1) {
+          throw new IllegalStateException(
+              String.format(
+                  "%s response type requested but multiple documents added, count=%s",
+                  responseType, documents.size()));
+        }
+        break;
+      case STATUS_ONLY:
+        if (!documents.isEmpty()) {
+          throw new IllegalStateException(
+              String.format(
+                  "%s response type requested but documents added, count=%s",
+                  responseType, documents.size()));
+        }
+        break;
+    }
+
+    if (!warnings.isEmpty()) {
+      cmdStatus.put(CommandStatus.WARNINGS, warnings);
+    }
+
+    // null out values that are empty, the CommandResult serialiser will ignore them when the JSON
```

Comment:
sidenote: this is fine, but there is also a Jackson annotation option for `@JsonInclude` to exclude both `null` and "empty" containers we can use.

     @JsonInclude(JsonInclude.Include.NON_EMPTY)

but it may be better not to rely on that here since that would also drop empty Strings as values.

Reply:
cool, thanks. I think there will be more work here. The initial phase is just to try to capture how it should work.